### PR TITLE
RATEST-301: Fix E2E Test: Patient feature workflow test

### PR DESCRIPTION
**Pull request by Irenyak1**
Ticket ID -> [RATEST-301](https://issues.openmrs.org/browse/RATEST-301)

Description -> E2E Test: Patient feature workflow

Legacy tests changed to BDD framwork

    FindPatientByNameTest,
    FindPatientRecordTest,
    RegisterUnidentifiedPatientTest ,
    DuplicatePatientRegisterTest,
    NamePatientAccentedLetterTest,
    MergePatientTest,
    UnidentifiedPatientKeyboardTest,
    UsingBackButtonInMergePatientTest

### CODE REVIEW

**Comment by Irenyak1** [comment]
Hello @kdaud , @sherrif10 , @mherman22 , @Muta-Jonathan could you please find some time to review my PR. Thanks.

File: qaframework-bdd-tests/src/test/resources/features/refapp-2.x/03-trackPatient/patient.feature [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+  	When the system loads Register a patient page
+  	And a user captures the patient details
+  	And the system alerts that similar patient found
+  	And a user clicks Confirm button
+    Then the system displays patient dashboard
+    
+  @selenium
+  @patient
+  Scenario: Name Patient Accented Letter
+  	Given a user clicks on Register a patient App from the home page
```
Comment (mherman22) [inline qaframework-bdd-tests/src/test/resources/features/refapp-2.x/03-trackPatient/patient.feature:67]: Thanks for the work @Irenyak1 please kindly look the indention in this file and correct it

**Review by mherman22 — CHANGES_REQUESTED** [review]

File: qaframework-bdd-tests/src/test/resources/features/refapp-2.x/03-trackPatient/patient.feature [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+  	When the system loads Register a patient page
+  	And a user clicks on the Unidentified Patient checkbox
+    And a user selects a gender
+    And a user clicks Confirm button
+    Then the system adds patient record into the patients table
+    
+  @selenium
+  @patient
+  Scenario: Unidentified Patient Keyboard
```
Comment (mherman22) [inline qaframework-bdd-tests/src/test/resources/features/refapp-2.x/03-trackPatient/patient.feature:39]: Whats the difference between this scenario and `Register Unidentified Patient`

Comment (Irenyak1) [reply qaframework-bdd-tests/src/test/resources/features/refapp-2.x/03-trackPatient/patient.feature:39]: @mherman22 they are very closely related, the only difference is that the former is testing registration of unidentified patient and the later is testing if it is possible to access the key board of unidentified patient.

File: qaframework-bdd-tests/src/test/resources/features/refapp-2.x/03-trackPatient/patient.feature (hunk 1 above)
Comment (Irenyak1) [reply qaframework-bdd-tests/src/test/resources/features/refapp-2.x/03-trackPatient/patient.feature:67]: @mherman22 sure thanks for pointing this out. I saw it and I am working on it.

File: qaframework-bdd-tests/src/test/resources/features/refapp-2.x/03-trackPatient/patient.feature [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+		When a user clicks on Data Management App from the home page
+		And  the system loads data management page
+		And a user clicks on Merge Patient Electronic Records app
+		And the system loads merge patients page
+		And a user captures IDs for patients to merge
+		And a user clicks Continue button
+		And a user clicks No button
+		And the system returns to capture patient ID section
+		And a user captures IDs for patients to merge
+		Then the system displays Continue button
```
Comment (kdaud) [inline qaframework-bdd-tests/src/test/resources/features/refapp-2.x/03-trackPatient/patient.feature:99]: @Irenyak1 go through the existing feature files, crosscheck the indention and align accordingly.

File: qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/PatientSteps.java [hunk 4]
```diff
@@ -0,0 +1,262 @@
...
+import io.cucumber.java.en.Given;
+import io.cucumber.java.en.Then;
+import io.cucumber.java.en.When;
+
+import static org.hamcrest.CoreMatchers.containsString;
+import static org.hamcrest.CoreMatchers.equalTo;
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertNotNull;
+import static org.junit.Assert.assertTrue;
```
Comment (kdaud) [inline qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/PatientSteps.java:24]: got through the import order of the existing files and align.

Comment (Irenyak1) [reply qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/PatientSteps.java:17]: @kdaud  I have attended to the suggestions. Could you make more reviews please.

File: qaframework-bdd-tests/src/test/resources/features/refapp-2.x/03-trackPatient/patient.feature [hunk 5]
```diff
@@ -1,99 +1,99 @@
...
+    And a user captures patient name with accented letter
+    And a user clicks Confirm button
+    Then the system displays patient dashboard
+
+  @selenium
+  @patient
+  Scenario: Merge Patients
+    Given a user creates two patients
+    When a user clicks on Data Management App from the home page
+    And  the system loads data management page
```
Comment (sherrif10) [inline qaframework-bdd-tests/src/test/resources/features/refapp-2.x/03-trackPatient/patient.feature:77]: Can you please correct the identation to be the same as the default identation

File: qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/PatientSteps.java [hunk 6]
```diff
@@ -0,0 +1,262 @@
...
+import org.openmrs.contrib.qaframework.page.DataManagementPage;
+import org.openmrs.contrib.qaframework.page.MergePatientsPage;
+import org.openmrs.contrib.qaframework.page.PatientVisitsDashboardPage;
+import org.openmrs.contrib.qaframework.page.RegistrationPage;
+
+public class PatientSteps extends Steps {
+	
+	private DataManagementPage dataManagementPage;
+	private MergePatientsPage mergePatientsPage;
+	private PatientVisitsDashboardPage VisitsDashboardPage;
```
Comment (kdaud) [inline qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/PatientSteps.java:39]: do you need to declare visitDashboardPage  variable?

File: qaframework-bdd-tests/src/test/resources/features/refapp-2.x/03-trackPatient/patient.feature (hunk 5 above)
Comment (Irenyak1) [reply qaframework-bdd-tests/src/test/resources/features/refapp-2.x/03-trackPatient/patient.feature:77]: > Can you please correct the identation to be the same as the default identation

Hi @sherrif10 please try refreshing on your side. I have just fixed the indentation.

File: qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/PatientSteps.java (hunk 6 above)
Comment (Irenyak1) [reply qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/PatientSteps.java:39]: > do you need to declare visitDashboardPage variable?

@kdaud Actually not. The mistake I made was to call it VisitDashboardPage.

File: qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/PatientSteps.java [hunk 7]
```diff
@@ -0,0 +1,260 @@
...
+		patient.city = "Kampala";
+		patient.state ="Central";
+		patient.country ="Uganda";
+		patient.postalCode ="235";
+		patient.phone ="0786987987";
+		registrationPage.enterPatient(patient);
+	}
+	
+	@And("the system alerts that similar patient found")
+	public void systemAllertsSimilarPatient() {
```
Comment (kdaud) [inline qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/PatientSteps.java:170]: take a close look at the method name

File: qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/PatientSteps.java [hunk 8]
```diff
@@ -0,0 +1,260 @@
...
+	
+	@And("a user clicks Confirm button")
+	public void clickOnConfirmButton() throws InterruptedException {
+  		dashboardPage = registrationPage.confirmPatient();
+	}
+	
+	@Then("the system adds patient record into the patients table")
+	public void systemAddsPatientRecord() {
+		assertThat(dashboardPage.getPatientGivenName(), equalTo("UNKNOWN"));
+		assertThat(dashboardPage.getPatientFamilyName(), equalTo("UNKNOWN"));
```
Comment (sherrif10) [inline qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/PatientSteps.java:139]: Is patientFamilyName similar to patientGivenName, could you use a different equalTo()

Comment (Irenyak1) [reply qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/PatientSteps.java:139]: > Is patientFamilyName similar to patientGivenName, could you use a different equalTo()

@sherrif10 no they are not similar, only that when an unidentified patient is registered, both the family name and given name are recorded as "UNKNOWN". So the patient will be "UNKNOWN UNKNOWN"

File: qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/PatientSteps.java (hunk 7 above)
Comment (Irenyak1) [reply qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/PatientSteps.java]: > take a close look at the method name

@kdaud  I have fixed it.

**Comment by Irenyak1** [comment]
@kdaud and @sherrif10  are there any more reviews on this issue? Could it be fit for merging into master?

File: qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/RunTest.java [hunk 9]
```diff
@@ -55,6 +56,7 @@
 		public static final String SELENIUM_VITALS = SELENIUM + " and "+ VITALS;
 		public static final String SELENIUM_CLINICAL_VISIT = SELENIUM + " and "+ CLINICAL_VISIT;
 		public static final String SELENIUM_ENCOUNTER = SELENIUM + " and "+ INPATIENT;
+		public static final String SELENIUM_PATIENT = SELENIUM + " and "+ PATIENT;
```
Comment (kdaud) [inline qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/RunTest.java:59]:
```suggestion
		public static final String SELENIUM_PATIENT = SELENIUM + " and " + PATIENT;
```

Comment (Irenyak1) [reply qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/RunTest.java:59]: @kdaud I think there's some inconsistency in this file. What would be the best way to concatenate these variables. So many do not have that space before the + sign. Are we planning to fix them for consistency?

Comment (kdaud) [reply qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/RunTest.java:59]: > Are we planning to fix them for consistency?

would be a good deal if you fix them all.

Comment (Irenyak1) [reply qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/RunTest.java:59]: > > Are we planning to fix them for consistency?
> 
> would be a good deal if you fix them all.

@kdaud would that be ok to fix them in this PR or I create a ticket for that?

Comment (kdaud) [reply qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/RunTest.java:59]: fix them in this PR.

Comment (Irenyak1) [reply qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/RunTest.java]: @kdaud  I have done the fix. Anymore reviews?

File: qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/PatientSteps.java [hunk 10]
```diff
@@ -0,0 +1,260 @@
...
+		findPatientPage = homePage.goToFindPatientRecord();
+	}
+	
+	@And("the system loads find patient page")
+	public void systemLoadsFindPatientPage() {
+		assertTrue(textExists("Find Patient Record"));
+	}
+	
+	@And("a user enters the patient's {string}")
+	public void findPatient(String name) {
```
Comment (kdaud) [inline qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/PatientSteps.java:71]: take a close look at the method name.

Comment (Irenyak1) [reply qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/PatientSteps.java]: @kdaud fixed

**Comment by Irenyak1** [comment]
@kdaud  and @mherman22  is there more work needed on this ticket or it can be merged?

**Comment by sherrif10** [comment]
@Irenyak1 report workflow should be functioning well. is your branch up to date

**Comment by mherman22** [comment]
> @mherman22 is there more work needed on this ticket

looks good to me after you update your branch with upstream changes.
over to you @kdaud and @sherrif10

**Comment by kdaud** [comment]
@Irenyak1 update your branch with master changes.

**Comment by kdaud** [comment]
@Irenyak1 I like the way you updated your branch with changes from master. How do you make it?

**Comment by Irenyak1** [comment]
Just wondering why the Patient Condition workflow is failing here https://github.com/openmrs/openmrs-contrib-qaframework/runs/5090971857?check_suite_focus=true#step:7:13375

> @Irenyak1 I like the way you updated your branch with changes from master. How do you make it?

@kdaud Is this real or it is a prophetic statement? :smile:

**Comment by kdaud** [comment]
> Just wondering why the Patient Condition workflow is failing here https://github.com/openmrs/openmrs-contrib-qaframework/runs/5090971857?check_suite_focus=true#step:7:13375

Have triggered it to run again, lets see how it goes.

**Comment by sherrif10** [comment]
> a prophetic statement?

It could be prophetic

**Comment by Irenyak1** [comment]
@sherrif10, @kdaud how can we have all firefox tests passing and yet the Patient condition workflow is failing!

**Comment by kdaud** [comment]
> how can we have all firefox tests passing and yet the Patient condition workflow is failing!

its possible for the individual workflow test to fail actions due to random failure but passes within All Firefox Tests

File: qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/PatientSteps.java [hunk 11]
```diff
@@ -0,0 +1,260 @@
...
+	
+	@And("a user searches for an existing patient by name")
+	public void searchPatientByName() {
+		findPatientPage.enterPatient(testPatient.givenName);
+		firstPatientIdentifier = findPatientPage.getFirstPatientIdentifier();
+		assertNotNull(firstPatientIdentifier);
+	}
+	
+	@And("a user clicks on first patient")
+	public void loadPatientDashboard() {
```
Comment (kdaud) [inline qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/PatientSteps.java:99]:
```suggestion
	public void loadPatientDashboardPage() {
```

File: qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/PatientSteps.java [hunk 12]
```diff
@@ -0,0 +1,260 @@
...
+		dashboardPage = findPatientPage.clickOnFirstPatient();
+	}
+	
+	@Then("the system loads the patient dashboard")
+	public void systemLoadsPatientDashboard() {
```
Comment (kdaud) [inline qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/PatientSteps.java:104]:
```suggestion
	public void systemLoadsPatientDashboardPage() {
```

File: qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/PatientSteps.java [hunk 13]
```diff
@@ -0,0 +1,260 @@
...
+	public void confirmMergePatientRecord() {
+		visitsDashboardPage = mergePatientsPage.clickOnContinue();
+	}
+	
+	@Then("the system loads patient visits dashboard")
+	public void systemLoadsPatientVisitsDashboard() {
+		assertThat(visitsDashboardPage.getPatientFamilyName(), equalTo(testPatient2.familyName));
+	}
+	
+	@And("a user clicks No button")
```
Comment (kdaud) [inline qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/PatientSteps.java:246]:
```suggestion
	@And("a user clicks on No button")
```

**Comment by kdaud** [comment]
@Irenyak1 have left some few comments to address!

**Comment by Irenyak1** [comment]
> @Irenyak1 have left some few comments to address!

@kdaud I have attended to all the reviews. You can make more.

**Comment by kdaud** [comment]
@Irenyak1 what do you think of the test name that will be displayed on the dashboard? We have already tests with the name Patient, is it possible to think of a unique name for this test?

**Comment by Irenyak1** [comment]
> @Irenyak1 what do you think of the test name that will be displayed on the dashboard? We have already tests with the name Patient, is it possible to think of a unique name for this test?

@kdaud  yeah I see that. Though many of the workflows have Patient with an additional name. What do you think could be the appropriate name? Can "Patient Record Management" work or "Patient Records" ? Advise.

**Comment by kdaud** [comment]
@Irenyak1 the test name is usually derived from what the test is accomplishing! So the test name you will come up should be in line with the context!

**Comment by Irenyak1** [comment]
> @Irenyak1 the test name is usually derived from what the test is accomplishing! So the test name you will come up should be in line with the context!

@kdaud the test is doing all this ;
    RegisterPatientTest,
    FindPatientByNameTest,
    _FindPatientRecordTest,
    RegisterUnidentifiedPatientTest _,
    _DuplicatePatientRegisterTest,
    NamePatientAccentedLetterTest,
    MergePatientTest,
    UnidentifiedPatientKeyboardTest,
    UsingBackButtonInMergePatientTest

It is not just a one clear workflow, it is doing many things. That's why I suggested what I suggested.

**Comment by mherman22** [comment]
> > @Irenyak1 the test name is usually derived from what the test is accomplishing! So the test name you will come up should be in line with the context!
> 
> @kdaud the test is doing all this ; RegisterPatientTest, FindPatientByNameTest, _FindPatientRecordTest, RegisterUnidentifiedPatientTest _, _DuplicatePatientRegisterTest, NamePatientAccentedLetterTest, MergePatientTest, UnidentifiedPatientKeyboardTest, UsingBackButtonInMergePatientTest
> 
> It is not just a one clear workflow, it is doing many things. That's why I suggested what I suggested.

how about `Patient Management`??



anyway am just passingby :walking:

**Comment by Irenyak1** [comment]
> how about `Patient Management`??

cc @kdaud  @sherrif10  what do you think?

**Comment by kdaud** [comment]
@Irenyak1 what name do you have on the table that you think would feet this workflow?

**Comment by Irenyak1** [comment]
> @Irenyak1 what name do you have on the table that you think would suit the workflow?

Patient Record Management

**Comment by kdaud** [comment]
how about removing the word Management?

**Comment by Irenyak1** [comment]
> how about removing the word `Management`?
@kdaud  since there's merging of records, I feel adding Management makes more meaning.

**Comment by kdaud** [comment]
> I feel adding Management makes more meaning.

initially all the workflows had been appended with the word `Management` and we decided to take the word off from them since its already captured within the title in their respective feature files.

**Comment by kdaud** [comment]
We can also add FindPatientRecordTest within this bdd test.

**Comment by Irenyak1** [comment]
> We can also add FindPatientRecordTest within this bdd test.

@kdaud It is one of the tests that are catered for here.

**Comment by kdaud** [comment]
> > We can also add FindPatientRecordTest within this bdd test.
> 
> @kdaud It is one of the tests that are catered for here.

Ooh I see!! The blind man told his deaf wife!

**Comment by Irenyak1** [comment]
@kdaud @sherrif10 could we have this merged?

**Review by kdaud — APPROVED** [review]
cc: @sherrif10

**Review by sherrif10 — APPROVED** [review]
Thanks @Irenyak1

**Pull request merged**